num.nix

```diff
@@ -153,4 +209,446 @@ rec {
+
+     Attempt to parse a string into an int. If parsing fails, throw an
+     exception.
+  */
+  parseInt = x:
+    let res = tryParseInt x;
+    in if res == null
+      then throw "std.num.parseInt: failed to parse"
+      else res;
+
+  /* tryParseFloat :: string -> maybe float
+
+     Attempt to parse a string into a float. If it fails and the string is still
+     parseable a valid JSON value, return null.
+  */
+  tryParseFloat = x:
+    let
+      # fromJSON aborts on invalid JSON values; check that it matches first
+      matches = builtins.match jsonNumberRE x != null;
+      res = builtins.fromJSON x;
+    in if matches && (builtins.isFloat res || builtins.isInt res)
+      then res
+      else null;
+
+  /* @partial
+     parseFloat :: string -> float
+
+     Attempt to parse a string into a float. If parsing fails, throw an
+     exception.
+  */
+  parseFloat = x:
+    let res = tryParseFloat x;
+    in if res == null
+      then throw "std.num.parseFloat: failed to parse"
+      else res;
+
+  /* toBaseDigits :: int -> int -> [int]
+
+     Convert an int to a list of digits in the given base. The most significant
+     digit is the first element of the list.
+
+     Note: this only works on positive numbers.
+  */
+  toBaseDigits = radix: x:
+    if x < 0
+      then throw "std.num.toBaseDigits: argument is negative"
+    else if radix == 1
+      then list.replicate x 1
+    else
+      let
+        go = xs: n:
+          if n < radix
+          then list.cons n xs
+          else
+            let qr = quotRem n radix;
+            in go (list.cons (qr._1) xs) qr._0;
+      in go [] x;
+
+  /* fromBaseDigits :: int -> [int] -> int
+
+     Convert a list of digits in the given base into an int. The most
+     significant digit is the first element of the list.
+  */
+  fromBaseDigits = radix: list.foldl' (acc: n: acc * radix + n) 0;
+
+  /* toHexString :: int -> string
+  */
+  toHexString = x:
+    let toHexDigit = string.index "0123456789ABCDEF";
+    in string.concatMap toHexDigit (toBaseDigits 16 x);
+
+  /* gcd :: int -> int -> int
+
+     Computes the greatest common divisor of two integers.
+  */
+  gcd = x: y:
+    let gcd' = a: b: if b == 0 then a else gcd' b (rem a b);
+    in gcd' (abs x) (abs y);
+
+  /* lcm :: int -> int -> int
+
+     Computes the least common multiple of two integers.
+  */
+  lcm = x: y:
+    if x == 0 || y == 0
+    then 0
+    else abs (quot x (gcd x y) * y);
+
+  /* clamp :: num -> num -> num
+
+     Clamp the value of the third argument to be between the first two
+     arguments, so that the result is lower-bounded by the first argument and
+     upper-bounded by the second.
+  */
+  clamp = lo: hi: x: min (max lo x) hi;
+
+  /* maxInt :: int
+
+     The highest representable positive integer.
+  */
+  maxInt = 9223372036854775807; # 2^63 - 1
+
+  /* minInt :: int
+
+     The lowest representable negative integer.
+  */
+  minInt = maxInt + 1; # relies on 2's complement representation
```

Comment:
can this just be hardcoded instead of relying on 2's complement? i mean it's unlikely that we'd deal with anything other than 2's complement, but if it's a hardcoded constant it's

1. more readable
2. more likely to constant fold (if marginally)

Reply:
It's not actually possible to write the literal; `-9223372036854775808` is not a literal unto itself, but the negation of `9223372036854775808`, which is not a valid literal in 64-bit 2's comp. Haskell has the same problem, thus GHC's `-XNegativeLiterals`.

If you prefer, I can write it as `(-9223372036854775807) - 1`.